torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit { namespace detail {
+
+// DynamicDAG is a DAG that dynamically maintains a topological order as
+// edges/vertices are added and removed.
+//
+// The implementation is based off of the following paper:
+// "A Dynamic Topsort Algorithm for Directed Acyclic Graphs"
+// by David Pearce and Paul Kelly
+// https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf
+
+template <typename T> struct Vertex;
+
+template <typename T>
+using vertex_set = std::unordered_set<Vertex<T>*>;
+
+template <typename T>
+using vertex_list = std::vector<Vertex<T>*>;
+
+template <typename T>
+using maybe_vertex_list = std::vector<at::optional<Vertex<T>*>>;
+
+template <typename T>
+struct Vertex {
+  std::vector<T> rdata; // stored in reverse order
+  size_t ord; // index in topological ordering. Is unique.
+  bool visited; // If this vertex has been visited
+  vertex_set<T> in_edges;
+  vertex_set<T> out_edges;
+
+  std::string toString();
+};
+
+template <typename T>
+struct DynamicDAG {
+  ~DynamicDAG();
+
+  Vertex<T>* newVertex(T datum);
+  std::pair<vertex_set<T>,vertex_set<T>> removeVertex(Vertex<T>* v);
+
+  void addEdge(Vertex<T>* producer, Vertex<T>* consumer);
+  void removeEdge(Vertex<T>* producer, Vertex<T>* consumer);
+  bool contractEdge(Vertex<T>* producer, Vertex<T>* consumer);
+
+  bool contractionProducesCycle(Vertex<T>* producer, Vertex<T>* consumer);
+  bool dfsForward(Vertex<T>* producer, size_t upper_bound, vertex_list<T>& visited);
+  void dfsBackward(Vertex<T>* producer, size_t lower_bound, vertex_list<T>& visited);
+  void clearVisited(vertex_list<T>& visited);
+
+  void reorder(vertex_list<T>& deltaF, vertex_list<T>& deltaB);
+  void sort(vertex_list<T>& delta);
+
+  void checkInvariants();
+
+  Vertex<T>* at(size_t ord) const { return vertices_.at(ord).value(); };
+  const maybe_vertex_list<T>& maybe_vertices() const { return vertices_; };
+
+  // Use for debugging. Don't call this often.
+  size_t numVertices() const;
+
+  std::string toString();
+
+ private:
+
+  // Store vertices indexed by their topological order.
+  // If a vertex v has ord 5, then it can be found at vertices_[5].
+  // There may be gaps in vertices_; this is to enable fast deletion.
```

Comment:
We don't care about insertion, because this assumes that we'll merge nodes into groups that execute atomically, right?

Reply:
Also, can't we simply store `at::optional<Vertex<T>>`? There's no reason for the heap indirection - this vector can provide you with all the lifetime management you need (it also improves data locality, because they're not scattered all over the heap).

Reply:
The algorithm reorders`Vertex<T>*` in the `std::vector<at::optional<Vertex<T>*>>`.

If it were `std::vector<at::optional<Vertex<T>>>` instead, reordering the Vertex<T> would involve copies of things larger than pointers, which is probably less performant. Let me know if that is not a correct statement.

Reply:
Yeah that makes sense. It would be great if we could make them `unique_ptr`s!